src/System.Net.Http/src/System/Net/Http/Managed/HttpConnectionKey.cs

```diff
@@ -6,35 +6,41 @@ namespace System.Net.Http
 {
     internal readonly struct HttpConnectionKey : IEquatable<HttpConnectionKey>
     {
-        public readonly bool UsingSSL;
         public readonly string Host;
         public readonly int Port;
+        public readonly string SslHostName;     // null if not SSL
 
-        public HttpConnectionKey(Uri uri)
+        public HttpConnectionKey(string host, int port, string sslHostName)
         {
-            UsingSSL =
-                HttpUtilities.IsSupportedNonSecureScheme(uri.Scheme) ? false :
-                HttpUtilities.IsSupportedSecureScheme(uri.Scheme) ? true :
-                throw new ArgumentException(SR.net_http_client_http_baseaddress_required, nameof(uri));
-
-            Host = uri.Host;
-            Port = uri.Port;
+            Host = host;
+            Port = port;
+            SslHostName = sslHostName;
         }
 
+        // The common cases here are SslHostName == null or SslHostName == Host
         public override int GetHashCode() =>
-            UsingSSL.GetHashCode() << 16 ^ Host.GetHashCode() ^ Port.GetHashCode();
+            SslHostName == null ? Host.GetHashCode() ^ Port.GetHashCode() :
```

Comment:
nit: in general I tend to put the repeated part, `Host.GetHashCode() ^ Port.GetHashCode()`, in a local but personal preference...